processor/k8sattributesprocessor/config.go

```diff
@@ -213,6 +213,9 @@ type FieldExtractConfig struct {
 	//       regex: JENKINS=(?P<value>[\w]+)
 	//
 	// this will add the `git.sha` and `ci.build` resource attributes.
+	// Deprecated: [v0.103.0] Use the `extract()` function of the transform processor instead.
+	// More information about how to replace regex parameter can be found under the following issue
+	// https://github.com/open-telemetry/opentelemetry-collector-contrib/issues/25128
 	Regex string `mapstructure:"regex"`
```

Comment:
We should follow our [end-user-facing deprecation process](https://github.com/open-telemetry/opentelemetry-collector/blob/main/CONTRIBUTING.md#end-user-impacting-changes) for this. Can you add a feature gate to the k8s attributes processor that, when enabled, returns an error in the config's `Validate` function if this field is set?

Reply:
Shouldn't it be deprecated first and added behind a gate in the following version? In the docs it states:

```
When deprecating a feature affecting end-users, consider first deprecating the feature in one version, then hiding it behind a [feature gate](https://github.com/open-telemetry/opentelemetry-collector/blob/6b5a3d08a96bfb41a5e121b34f592a1d5c6e0435/service/featuregate/) in a subsequent version, and eventually removing it after yet another version.
```

Please correct me if I do not understand it right

Reply:
That's a fair reading of that statement, those docs should probably be reworded a bit. The intended process is that you leave the feature gate "disabled" such that users can opt-in to the new functionality for the first step, then make the new functionality opt-out, then finally make it required and remove the feature gate.

Reply:
> The intended process is that you add the feature gate but leave off by default such that users can opt-in to the new functionality for the first step

A little more context if it's helpful, a feature gate at `Alpha` stability level is disabled by default, `Beta` is enabled by default, and `Stable` can't be disabled. ([Doc reference](https://github.com/open-telemetry/opentelemetry-collector/blob/main/featuregate/README.md#feature-lifecycle))

So if you add a feature gate at `Alpha` stability, it will behave as @evan-bradley has suggested, and then the following change is simply to update its stability level. 👍

Reply:
In my mind change/removal of certain functionality is usually hidden behind a feature gate, not removing/deprecating a parameter. I do not think adding a feature gate for a parameter is a standard way to do this (let's just look at Kubernetes deprecating parameters in their APIs). As a deprecated parameter, you can have a warning about it, but you should still be able to use it. The next step (after few versions) is removing it.

I am confused what do you expect to be hidden behind the feature flag

Thanks for the answers.

Reply:
A warning is a valid option as well. [Our docs](https://github.com/open-telemetry/opentelemetry-collector/blob/main/CONTRIBUTING.md#beta-components) also cover how to handle deprecations in a case like this.

The one upside of the feature gate is that we can turn off the ability to use the config field unless you explicitly disable the feature gate, and push users to migrate early. So long as we're printing a warning with a clear migration path well in advance I think both options could work. We could also go with a warning log for now and add a feature gate later, since the deprecation process is intended to be so long.

@dmitryax @TylerHelmuth as the code owners involved in the original issue, do you have a preference?

Reply:
In the OTel Collector, we typically use feature gates to manage user-facing changes, including configuration changes.  Having the feature gate allows users to hit a failure, disable the gate instead of updating config, and then work on updating config before the next upgrade.

The feature gate essentially gives the end user 1 more chance to fix the problem.  While warnings are nice, in reality a lot of people don't see them.  With the feature gate you can force the collector to error with a very clear message about how to handle the error or disable the gate, and then go make changes.

Reply:
Adapted, thanks for the discussion!